src/lang/std/engineConnection.ts

```diff
@@ -2089,4 +2089,25 @@ export class EngineCommandManager extends EventTarget {
       },
     })
   }
+
+  // Some "objects" have the same source range, such as sketch_mode_start and start_path.
+  // So when passing a range, we need to also specify the command type
+  mapRangeToObjectId(
+    range: SourceRange,
+    commandTypeToTarget: string
+  ): string | undefined {
+    const values = Object.entries(this.artifactMap)
+    for (const [id, data] of values) {
+      if (data.type !== 'result') continue
+
+      // Our range selection seems to just select the cursor position, so either
+      // of these can be right...
+      if (
+        (data.range[0] === range[0] || data.range[1] === range[1]) &&
+        data.commandType === commandTypeToTarget
+      )
+        return id
+    }
+    return undefined
+  }
```

Comment:
There's very similar logic in `selections.ts`
```typescript
function codeToIdSelections(
  codeBasedSelections: Selection[]
): SelectionToEngine[] {
  return codeBasedSelections
    .flatMap(({ type, range, ...rest }): null | SelectionToEngine[] => {
      // TODO #868: loops over all artifacts will become inefficient at a large scale
      const entriesWithOverlap = Object.entries(
        engineCommandManager.artifactMap || {}
      )
        .map(([id, artifact]) => {
          return artifact.range && isOverlap(artifact.range, range)
            ? {
                artifact,
                selection: { type, range, ...rest },
                id,
              }
            : false
        })
        .filter(Boolean)
      let bestCandidate
      // ...
```
We might consider sharing logic between the two?

Reply:
maybe not, they're not 1:1, I'll leave it up to you.

Reply:
We need to eventually rework how we get ids.

You're right these are similar but the issue with codeToIdSelections is sometimes there is code ranges that refers to two different objects. For mapRangeToObjectId, this can't happen because of the targetCommandType.

I would rather leave it and rework our code to remove both of these routines in the future.

Reply:
Yeah cool, relevant issue for artifactMap btw
https://github.com/KittyCAD/modeling-app/issues/868